pkg/snapshot/file/publisher.go

```diff
@@ -0,0 +1,297 @@
+		&snapt.TableStateNode,
+		&snapt.TableStorageNode,
+	}
+)
+
+const logInterval = 1 * time.Minute
+
+type publisher struct {
+	dir     string // dir containing output files
+	writers fileWriters
+
+	nodeInfo nodeinfo.Info
+
+	startTime          time.Time
+	currBatchSize      uint
+	stateNodeCounter   uint64
+	storageNodeCounter uint64
+	codeNodeCounter    uint64
+	txCounter          uint32
+}
+
+type fileWriter struct {
+	*csv.Writer
+}
+
+// fileWriters wraps the file writers for each output table
+type fileWriters map[*snapt.Table]fileWriter
+
+type fileTx struct {
+	fileWriters
+	// index uint
+}
+
+func (tx fileWriters) Commit() error {
+	for _, w := range tx {
+		w.Flush()
+		if err := w.Error(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (fileWriters) Rollback() error { return nil } // TODO: delete the file?
+
+func newFileWriter(path string) (ret fileWriter, err error) {
+	file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
+	if err != nil {
+		return
+	}
+	buf := file //bufio.NewWriter(file)
+	ret = fileWriter{csv.NewWriter(buf)}
+	return
+}
+
+func (tx fileWriters) write(tbl *snapt.Table, args ...interface{}) error {
+	row := tbl.ToCsvRow(args...)
+	return tx[tbl].Write(row)
+}
+
+func makeFileWriters(dir string, tables []*snapt.Table) (fileWriters, error) {
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+	writers := map[*snapt.Table]fileWriter{}
+	for _, tbl := range tables {
+		w, err := newFileWriter(TableFile(dir, tbl.Name))
+		if err != nil {
+			return nil, err
+		}
+		writers[tbl] = w
+	}
+	return writers, nil
+}
+
+// NewPublisher creates a publisher which writes to per-table CSV files which can be imported
+// with the Postgres COPY command.
+// The output directory will be created if it does not exist.
+func NewPublisher(path string, node nodeinfo.Info) (*publisher, error) {
+	if err := os.MkdirAll(path, 0755); err != nil {
+		return nil, err
+	}
+	writers, err := makeFileWriters(path, perBlockTables)
+	if err != nil {
+		return nil, err
+	}
+	pub := &publisher{
+		writers:   writers,
+		dir:       path,
+		nodeInfo:  node,
+		startTime: time.Now(),
+	}
+	go pub.logNodeCounters()
+	return pub, nil
+}
+
+func TableFile(dir, name string) string { return filepath.Join(dir, name+".csv") }
+
+func (p *publisher) txDir(index uint32) string {
+	return filepath.Join(p.dir, fmt.Sprintf("%010d", index))
+}
+
+func (p *publisher) BeginTx() (snapt.Tx, error) {
+	index := atomic.LoadUint32(&p.txCounter)
+	dir := p.txDir(index)
+	writers, err := makeFileWriters(dir, perNodeTables)
+	if err != nil {
+		return nil, err
+	}
+
+	atomic.AddUint32(&p.txCounter, 1)
+	return fileTx{writers}, nil
+}
+
+// PublishRaw derives a cid from raw bytes and provided codec and multihash type, and writes it to the db tx
+// returns the CID and blockstore prefixed multihash key
+func (tx fileWriters) publishRaw(codec uint64, raw []byte) (cid, prefixedKey string, err error) {
```

Comment:
Yeah I'm not sure how I feel about this fileWriters tx approach but I might be missing something.

Couldn't we instead, at the beginning of processing a new block+it's state, spin up a goroutine worker for writing each type of statement to their respective destination files.

The worker's state would be the file handle, the stmt they write, a work chan, and a quit/close chan. Quit/close chan would be shared across all workers.

And then we could feed them work (the specific args of their statements) over their work chans as we iterate the data structures. Concurrent access to the file handles would be synced by the gating/funnel action of the worker and chan. And when the trie iteration is complete we would `close(quitChan)` and that would close all the workers which would have a defer() that flushes their file buffers and closes their file handles.

Reply:
My intention here was to avoid having to sync the output at all, since each tx uses a separate file. If we want to have a single file for each table (or any mapping of workers to files that's not 1 to 1), channels would be the way to go, but do you expect any performance benefit in doing that? It seems like we could get the same result by just concatenating the files in a separate stage. Could be useful to benchmark each approach.

Reply:
Thanks Roy, that makes sense! I think you are right about concatenating in a separate stage being better. And we can come back to this later if we want to, but shouldn't worry about that benchmark right now.